### PR TITLE
Fix #3004 and simplify exhaustivity check

### LINKED ISSUE #3004: Custom unapply methods can produce misleading error messages
As a simple example, the following code:
```scala
object O {
         sealed trait Fruit
         object Apple extends Fruit
         object Banana extends Fruit
         sealed class C(f1: Fruit, f2: Fruit)
         object C {
           def unapply(c: C): Some[Banana.type] = Some(Banana)
         }
         def m(c: C) = c match { case C(b) => b }
}
```
produces the following message:
```
-- [E028] Pattern Match Exhaustivity Warning: <console>:12:16 ------------------
12 |  def m(c: C) = c match { case C(b) => b }
   |                ^
   |                match may not be exhaustive.
   |
   |                It would fail on: C(Apple, _)
```
The pattern match is, in fact, exhaustive and the message is somewhat puzzling.

The underlying issue is the interaction between this code in the `project` method:
https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/patmat/Space.scala#L430-L431
and the places in rest of the code which "project" `Typ(tp)` as `Kon(tp, signature(tp).map(Typ(_))`:
https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/patmat/Space.scala#L267
When a `Kon` space resulting from projecting `C.unapply` is subtracted from `Typ(C)`, "inconsistent" `Kon` spaces are created, which then in turn causes this issue.

/cc @liufengyun

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #3004 and simplify exhaustivity check

### CODE REVIEW

**Comment by liufengyun** [comment]
Review @AleksanderBG ?

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 1]
```diff
@@ -641,6 +628,15 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
 
   /** Display spaces */
   def show(s: Space): String = {
+    def params(tp: Type): List[Type] = {
```
Comment (AleksanderBG) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: This function seems to be mostly doing unnecessary work - the actual types returned are never used, and the only thing that matters is the length of the returned list.

**Comment by AleksanderBG** [comment]
@liufengyun - I've tried to review what I can, and apart from that single nitpick everything looks good to me. I'm a bit out of my depth when it comes to verifying that the new implementation of `signature` is correct, though. I must say that unifying `Kon` and `Fun` spaces seems like a very good approach - the overall algorithm now looks much cleaner.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala (hunk 1 above)
Comment (AleksanderBG) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: To clarify, I've checked that the length of `.firstParamTypes` does not actually change - all tests still pass if this function returns `ktor.fistParamTypes`.

File: tests/patmat/t9232.check [hunk 2]
```diff
@@ -1 +1,3 @@
-13: Pattern Match Exhaustivity: Node2(), Node1(Foo(_))
+13: Pattern Match Exhaustivity: Node2()
+17: Pattern Match Exhaustivity: Node2(), Node1(Foo(Nil)), Node1(Foo(List(_, _*)))
```
Comment (smarter) [inline tests/patmat/t9232.check]: Should `List(_, _*)` be `List(_, _: _*)` ?

**Comment by liufengyun** [comment]
Thanks @AleksanderBG . The `signature` is stolen from the new patternMatcher, which follows the specification closely.

**Comment by smarter** [comment]
> The signature is stolen from the new patternMatcher, which follows the specification closely.

Would be nice to add a comment pointing to the corresponding code in the pattern matcher. Even nicer if we could share some code but I don't know if that's possible here :).

**Comment by liufengyun** [comment]
test performance please

**Pull request merged**